mozilla/socorro
# fixes bug 1262252 - Crash data pure implementation

Again, sorry for the size of this. I'll try to break it down. 

## The general gist

### Implementations
It used to be that crash data got looked up via the middleware_app and ultimately it would be `socorro.external.crash_data_base.CrashDataBase` that would do the wrapping of the `S3BotoCrashStorage` to pull down raw dumps, raw crashes, processed crashes from S3. That class would also do too much. It would listen for boto StorageError exceptions and if that happens it would put the crash ID onto the priority jobs queue in rabbitmq.

What I've done is break that apart. And simplified. Instead of re-writing `socorro.external.crash_data_base.CrashDataBase` I created a fresh new class called `socorro.external.boto.crash_data.SimplifiedCrashData`. First of all, it's put in `boto` folder which is very explicit. Second, if this works out and goes into production I might rename it back to plain `CrashData` once I've cleaned up old legacy code. 

Then there's a another brand new class that deals specifically with putting jobs into the priority queue. That's all it does. 

[Deleting more old legacy code will be done later](https://bugzilla.mozilla.org/show_bug.cgi?id=1299465).

### Webapp
[Before it used be weird](https://github.com/mozilla/socorro/blob/344979da66b185fc61ba660e9848591bf6a1ef14/webapp-django/crashstats/crashstats/views.py#L1359-L1378). When loading the report index it would first try to load the [unredacted] processed crash *then* later try to load the raw crash. With some magical juggling of types of exceptions from the middleware it would either display a page that says "Crash Not Found", "Crash Too Old" or "Crash Pending, fetching it now". 

I just vastly simplified that. So **first** it loads the raw crash. If that fails (e.g. `CrashIDNotFound`) it plain and simply says "Crash Not Found". Secondly, when it loads the processed crash, if it's not found it puts the crash ID on the priority job AND returns a "Crash Pending" template (error code 200) and that would load some JavaScript code that checks `/api/ProcessedCrash/?crash_id=:crash_id` repeatedly until it exists and then reloads the page. 
[I hope to improve the "Crash Not Found" error page later](https://bugzilla.mozilla.org/show_bug.cgi?id=1299460).

## How to test it
Load up a report index page (using stage data). Go into the AWS management console and go to the `org.allizom.crash-stats.rhelmer` bucket and click into the [v1/processed_crash directory](https://console.aws.amazon.com/s3/home?region=us-west-2#&bucket=org.allizom.crash-stats.rhelmer-test.crashes&prefix=v1/processed_crash/). Look up the crash ID (that you loaded in the report index) and find it in AWS with the "Search by prefix" search. Once found, simply delete the processed crash json file. Then **restart** your webapp (to clear the LocMemCache) and refresh the report index URL you were on. (Make sure you have the Web Console open!) It should now load the "pending template" and it should just work :)

## The mocking
The major reason this PR is so HUGE is because `crashstats/crashstats/tests/test_views.py` does a lot of tests on top of `report_index`. Every single line of mocking (implementation directly instead of middleware JSON) had to be re-written. As I was doing that I refactored a couple of places where it had a big hardcoded python dict and use the `_SAMPLE_META` or `_SAMPLE_UNREDACED` dicts.
